File: day2/main.py
import random

points = {
    "X": 1,  # Rock
    "Y": 2,  # Paper
    "Z": 3,  # Scissors
    # "A"  Opponent Rock
    # "B"  Opponent Paper
    # "C"  Opponent Scissors
    "Lost": 0,
    "Draw": 3,
    "Won": 6,
}


def get_winner(p1, p2):
    if (p1 == "C" and p2 == "X") or (p1 == "B" and p2 == "Z") or (p1 == "A" and p2 == "Y"):
        return (1, p2)  # Player wins
    elif (p1 == "A" and p2 == "X") or (p1 == "B" and p2 == "Y") or (p1 == "C" and p2 == "Z"):
        return (-1, p2)  # Draw
    else:
        return (0, p2)  # Player loses


def get_points(played):
    (winner, chosen) = get_winner(played[0], played[2])
    win_score = 0
    match winner:
        case -1:
            win_score = points["Draw"]
        case 0:
            win_score = points["Lost"]
        case 1:
            win_score = points["Won"]
    return win_score + points[chosen]


def choose_move(game):
    p1 = game[0]
    p2 = game[2]
    draw = {
        "A": "X",
        "B": "Y",
        "C": "Z"
    }

    win = {
        "A": "Y",
        "B": "Z",
        "C": "X",
    }

    lose = {
        "A": "Z",
        "B": "X",
        "C": "Y"
    }
    match p2:
        case "X":  # Lose
            return f"{p1} {lose[p1]}"
        case "Y":  # Draw
            return f"{p1} {draw[p1]}"
        case "Z":  # Win
            return f"{p1} {win[p1]}"


with open("./input", "r") as f:
    f = f.readlines()
    moves = list(map(choose_move, f))
    player_points = list(map(get_points, moves))
    print(sum(player_points))

File: day2/Main.hs
module Main where

import Data.List.Split (splitOn)
import System.Environment (getArgs)

getPlays :: String -> [[String]]
getPlays file = map (splitOn " ") (lines file)

getWinner :: String -> String -> (Int, String)
getWinner p1 p2 = case (p1, p2) of
  ("C", "X") -> wins
  ("B", "Z") -> wins
  ("A", "Y") -> wins
  ("A", "X") -> draws
  ("B", "Y") -> draws
  ("C", "Z") -> draws
  _ -> loses
  where
    wins = (1, p2)
    draws = (-1, p2)
    loses = (0, p2)

getPoints :: [String] -> Int
getPoints play = case chosen of
  "X" -> 1 + playPoint
  "Y" -> 2 + playPoint
  "Z" -> 3 + playPoint
  where
    (winner, chosen) = getWinner (head play) (play !! 1)
    playPoint = case winner of
      -1 -> 3 -- won
      0 -> 0 -- lost
      1 -> 6 -- drew

-- only required for part 2
chooseMove :: [String] -> [String]
chooseMove [p1, p2] =
  [p1]
    <> [ case p2 of
           -- should lose
           "X" -> case p1 of
             "A" -> "Z"
             "B" -> "X"
             "C" -> "Y"
           -- should draw
           "Y" -> case p1 of
             "A" -> "X"
             "B" -> "Y"
             "C" -> "Z"
           -- should win
           "Z" -> case p1 of
             "A" -> "Y"
             "B" -> "Z"
             "C" -> "X"
       ]

main :: IO ()
main = do
  args <- getArgs
  file <- readFile (head args)
  let plays = chooseMove <$> getPlays file
  let points = fmap getPoints plays
  print . sum $ points
